runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -851,27 +853,53 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
                                        TR::TreeTop* indirectAccessTreeTop,
                                        bool needNullCheck, bool isUnsafeGet,
                                        bool conversionNeeded,
+                                       bool arrayCheckNeeded, bool arrayBlockNeeded,
+                                       bool checksNeeded,
                                        TR::Node* orderedCallNode = NULL)
    {
+   // There are 4 possible cases determining which checks and blocks should be added to the generated IL trees:
+   // 1.) checksNeeded && arrayCheckNeeded && arrayBlockNeeded
+   //     (i.e.: (offheap AND object type is unknown at compile time) OR conversionNeeded)
+   //     - checks: NULLCHK, ArrayCHK, LowtagCHK
+   //     - blocks: arrayDirectAccess, directAccess, indirectAccess
+   // 2.) checksNeeded && !arrayCheckNeeded && !arrayBlockNeeded
+   //     (i.e.: gencon AND object type is unknown at compile time AND !conversionNeeded)
+   //     - checks: LowtagCHK, NULLCHK, ClassCHK
+   //     - blocks: directAccess, indirectAccess
+   // 3.) !checksNeeded && !arrayCheckNeeded && arrayBlockNeeded
+   //     (i.e.: offheap AND object is known to be array at compile time)
+   //     - checks: NULLCHK
+   //     - blocks: arrayDirectAccess, directAccess
+   // 4.) !checksNeeded && !arrayCheckNeeded && !arrayBlockNeeded
+   //     (i.e.: (offheap AND object is known to be non-array object at compile) time OR gencon)
```

Comment:
I went with `arrayTestNeeded` and `typeTestsNeeded` to replace `arrayCheckNeeded` and `checksNeeded` for now. I am a little hesitant to rename `typeTestsNeeded` to `lowTagTestNeeded`, just because `typeTestsNeeded == true` is a requirement for `arrayTestNeeded` to evaluate to `true`. The intent there is that we first need to check if ANY tests at all (which could include a lowtag test, an array test, and/or a class test) are necessary, and _then_ determine if an array test specifically is needed. I worry that naming it `lowTagTestNeeded` may not get that across clearly. 

Same with the concern about the parameter names being too descriptive of the implementation details. I agree that it's not ideal, but I'm not sure that there's another way to communicate what those variables do efficiently...at least not with how they are currently used.